README.md
<h1 align="center">
    <br>
    Av1an
    </br>
</h1>

<h2 align="center">A cross-platform framework to streamline encoding</h2>

![alt text](https://cdn.discordapp.com/attachments/696849974666985494/774368268860915732/av1an_pick2.png)

<h4 align="center">
<a href="https://discord.gg/Ar8MvJh"><img src="https://discordapp.com/api/guilds/696849974230515794/embed.png" alt="Discord server" /></a>
<img src="https://github.com/master-of-zen/Av1an/workflows/tests/badge.svg">
<a href="https://codeclimate.com/github/master-of-zen/Av1an/maintainability"><img src="https://api.codeclimate.com/v1/badges/41ea7ad221dcdad3fe8d/maintainability" />
<img src="https://app.codacy.com/manual/Grenight/Av1an?utm_source=github.com&utm_medium=referral&utm_content=master-of-zen/Av1an&utm_campaign=Badge_Grade_Dashboard"></a>
<a href="https://www.codacy.com/manual/Grenight/Av1an?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=master-of-zen/Av1an&amp;utm_campaign=Badge_Grade"><img src="https://api.codacy.com/project/badge/Grade/4632dbb2f6f34ad199142c01a3eb2aaf"/></a>
</h4>
<h2 align="center">Easy, Fast, Efficient and Feature Rich</h2>

An easy way to start using VVC / AV1 / HEVC / H264 / VP9 / VP8 encoding. AOM, RAV1E, SVT-AV1, SVT-VP9, VPX, x265, x264, VTM(Experimental) are supported.

<h2 align="center">Quick Install</h2>

     pip install git+https://github.com/Thomasedv/Av1an

See [Install](#Install) section below for more details. Av1an relies on external program to function.

This is a fork of [this](https://github.com/master-of-zen/Av1an) repository. You may have been looking for that one in the first place.

<h2 align="center">Examples</h2>

Example with default parameters:

    av1an -i input

Simple target quality code:

    av1an -i input --target_quality 95 --vmaf_path "vmaf_v0.6.1.json"


With your own parameters:

    av1an -i input -enc aom -v " --cpu-used=3 --end-usage=q --cq-level=30 --threads=8 " -w 10
    --split_method aom_keyframes --target_quality 95 --vmaf_path "vmaf_v0.6.1.json" 
    -min_q 20 -max_q 60 -ff "-vf scale=-1:1080" -a "-c:a libopus -ac 2 -b:a 192k" 
    -s scenes.csv -log my_log -o output

<h2 align="center">Usage</h2>

    -i   --input            Input file(s), or Vapoursynth (.py,.vpy) script 
                            (relative or absolute path)

    -o   --output_file      Name/Path for output file (Default: (input file name)_(encoder).mkv)
                            Output file ending is always `.mkv`

    -enc --encoder          Encoder to use
                            (`aom`,`rav1e`,`svt_av1`,`svt_vp9`,`vpx`,`x265`, `x264`,`vvc`)
                            Default: aom
                            Example: -enc rav1e

    -v   --video_params     Encoder settings flags (If not set, will be used default parameters.)
                            Must be inside ' ' or " "

    -p   --passes           Set number of passes for encoding
                            (Default: AOMENC: 2, rav1e: 1, SVT-AV1: 1, SVT-VP9: 1,
                            VPX: 2, x265: 1, x264: 1, VVC:1)

    -w   --workers          Override number of workers.

    -r   --resume           If encode was stopped/quit resumes encode with saving all progress.
                            Resuming automatically skips scenedetection, audio encoding/copy,
                            splitting, so resuming only possible after actual encoding is started.
                            Temp folder must be present to resume.

    --no_check              Skip checking numbers of frames for source and encoded chunks.
                            Needed if framerate changes to avoid console spam.
                            By default, any differences in frames of encoded files will be reported.

    --keep                  Doesn't delete temporary folders after encode has finished.

    -q --quiet              Do not print tqdm to the terminal.

    -log --logging          Path to .log file(By default created in temp folder)

    --temp                  Set path for the temporary folder. Default: .temp

    --mkvmerge              Use mkvmerge for concatenating instead of FFmpeg.
                            Use when concatenation fails.

    -c  --config            Save/Read config file with encoder, encoder parameters,
                            FFmpeg and audio settings.
                            Options provided to cli overwrite config values.
                            All options except in/out/VMAF/log/temp/config paths are saved.

    --webm                  Outputs webm file.
                            Use only if you're sure the source video and audio are compatible.

    --priority              Sets process priority for encoding and frame probes. (Windows only!)
                            (`normal`, `low`, `idle`, `high`)
                            Default: normal
                            

<h3 align="center">FFmpeg options</h3>

    -a   --audio_params     FFmpeg audio settings (Default: copy audio from source to output)
                            Example: -a '-c:a libopus -b:a  64k'

    -ff  --ffmpeg           FFmpeg options video options. 
                            Applied to each encoding segment individually.
                            (Warning: Cropping doesn't work with Target VMAF mode 
                            without specifying it in --vmaf_filter)
                            Example:
                            --ff " -vf scale=320:240 "

    -fmt --pix_format       Setting custom pixel/bit format for piping
                            (Default: 'yuv420p10le')
                            Options should be adjusted accordingly, based on the encoder.

<h3 align="center">Segmenting</h3>

    --split_method          Method used for generating splits.(Default: PySceneDetect)
                            Options: `pyscene`, `aom_keyframes`, `none`
                            `pyscene` - PyScenedetect, content based scenedetection
                            with threshold.
                            `aom_keyframes` - using stat file of 1 pass of aomenc encode
                            to get exact place where encoder will place new keyframes.
                            (Keep in mind that speed also depends on set aomenc parameters)
                            `ffmpeg` - Uses FFmpeg built in content based scene detection
                            with threshold. Slower and less precise than pyscene but requires
                            fewer dependencies.
                            `none` -  skips scenedetection. Useful for splitting by time

    -cm  --chunk_method     Determine the method in which chunks are made for encoding.
                            By default the best method is selected automatically in this order:
                            vs_ffms2 > vs_lsmash > hybrid.
                            vs_ffms2 or vs_lsmash are recommended.
                            ['hybrid'(default), 'select', 'vs_ffms2', 'vs_lsmash']


    -tr  --threshold        PySceneDetect threshold for scene detection Default: 35

    -s   --scenes           Path to file with scenes timestamps.
                            If the file doesn't exist, a new file will be generated 
                            in the current folder.
                            First run to generate stamps, all next reuse it.
                            Example: "-s scenes.csv"

    -xs  --extra_split      Adding extra splits if frame distance between splits bigger than the
                            given value. Pair with none for time based splitting or with any
                            other splitting method to break up massive scenes.
                            Example: 1000 frames video with a single scene,
                            -xs 200 will add splits at 200,400,600,800.

    --min_scene_len         Specifies the minimum number of frames in each split.

<h3 align="center">Target Quality</h3>

    --target_quality        Quality value to target.
                            VMAF used as substructure for algorithms.
                            Supported in all encoders supported by Av1an except for VVC.
                            Best works in range 85-97.
                            When using this mode specify full encoding options.
                            Encoding options must include quantizer based mode,
                            and some quantizer option provided. (This value will be replaced)
                            `--crf`,`--cq-level`,`--quantizer` etc

    --target_quality_method Type of algorithm for use.
                            Options: per_shot, per_frame.
                            Per frame is only supported in SVT-AV1.

    --min_q, --max_q        Min,Max Q values limits
                            If not set by the user, the default for encoder range will be used.

    --vmaf                  Calculate VMAF after encode is done and make a plot.

    --vmaf_plots            Make plots for target quality search decisions
                            (Exception: early skips)
                            Saved in the temp folder by default.

    --vmaf_path             Custom path to libvmaf models.
                            example: --vmaf_path "vmaf_v0.6.1.pkl"
                            Recomended to place both files in encoding folder
                            (`vmaf_v0.6.1.pkl` and `vmaf_v0.6.1.pkl.model`)
                            (Required if VMAF calculation doesn't work by default)

    --vmaf_res              Resolution scaling for VMAF calculation,
                            vmaf_v0.6.1.pkl is 1920x1080 (by default),
                            vmaf_4k_v0.6.1.pkl is 3840x2160 (don't forget about vmaf_path)

    --probes                Number of probes for interpolation.
                            1 and 2 probes have special cases to try to work with few data points.
                            The optimal level is 4-6 probes. Default: 4

    --vmaf_filter           Filter used for VMAF calculation. The passed format is filter_complex.
                            So if crop filter used ` -ff " -vf crop=200:1000:0:0 "`
                            `--vmaf_filter` must be : ` --vmaf_filter "crop=200:1000:0:0"`

    --probing_rate          Setting rate for VMAF probes (Every N frame used in probe, Default: 4)

    --n_threads             Limit number of threads that are used for VMAF calculation
                            Example: --n_threads 12
                            (Required if VMAF calculation gives error on high core counts)

<h2 align="center">Main Features</h2>

**Splitting video by scenes for parallel encoding** because AV1 encoders are currently not very good at multithreading and encoding is limited to a very limited number of threads.

-   [PySceneDetect](https://pyscenedetect.readthedocs.io/en/latest/) used for splitting video by scenes and running multiple encoders.
-   [Vapoursynth](http://www.vapoursynth.com) script input support.
-   Fastest way to encode AV1 without losing quality, as fast as many CPU cores you have :).
-   Target Quality mode. Targeting end result reference visual quality. VMAF used as a substructure
-   Resuming encoding without loss of encoded progress.
-   Simple and clean console look.
-   Automatic detection of the number of workers the host can handle.
-   Building encoding queue with bigger files first, minimizing waiting for the last scene to encode.
-   Both video and audio transcoding with FFmpeg.
-   Logging of the progress of all encoders.
-   You can process a folder of files by using the folder as input, by ending the input name with `\ ` or `/`: 

    - `av1an -i folder\ ` or `av1an -i folder/ `

    - The same can be done for temporary folders (where av1an saves logs and temporary files for each input file)
    - `av1an -i folder\  --temp temp_folder\ ` or `av1an -i folder/ --temp temp_folder/`


# Install

-   Prerequisites:
    -   [Install Python3.7+](https://www.python.org/downloads/) <br>
        When installing under Windows, select the option `add Python to PATH` in the installer
    -   [Install FFmpeg](https://ffmpeg.org/download.html)
    -   Recommended to install vapoursynth with lsmash for faster and better processing

-   Encoder of choice:
    -   [Install AOMENC](https://aomedia.googlesource.com/aom/)
    -   [Install rav1e](https://github.com/xiph/rav1e)
    -   [Install SVT-AV1](https://gitlab.com/AOMediaCodec/SVT-AV1)
    -   [Install SVT-VP9](https://github.com/OpenVisualCloud/SVT-VP9)
    -   [Install vpx](https://chromium.googlesource.com/webm/libvpx/) VP9, VP8 encoding
    -   [Install VTM](https://vcgit.hhi.fraunhofer.de/jvet/VVCSoftware_VTM) VVC encoding test model

-   Optional :

    -   [Vapoursynth](http://www.vapoursynth.com/)
    -   [ffms2](https://github.com/FFMS/ffms2)
    -   [lsmash](https://github.com/VFR-maniac/L-SMASH-Works)
    -   [mkvmerge](https://mkvtoolnix.download/)

-   Manual Av1an install: 
    - With pip
      
        `pip install git+https://github.com/Thomasedv/Av1an`        

    - Clone Repo or Download from Releases
        
        `python setup.py install`


-   Also:
    On Ubuntu systems, the packages `python3-opencv` and `libsm6` are required

### Support the developer (The original creator: master-of-zen)

Bitcoin - 1GTRkvV4KdSaRyFDYTpZckPKQCoWbWkJV1


CHANGELOG.md
### 6
- ~1.7x faster probes for svt-av1
- ~2x faster probes for aomenc
-  Changed rav1e settings, disable rav1e scene detection
- Temporally removed VVC support until it's 100% ready and working
- Added overwrite promt


### 5
- Fixed fatal errors with ffms2/lsmash
- Added vmaf validation on each time when VMAF initialized
- Fix running not required frame probe
- Chunk restarting
- Fixed ffmpeg segmenting
- aomenc target quality probes to be 8 bit

### 4
- Refactored Args to Project class
- Removed dead Rust code
- Default encoder settings changed
- Better Vapousynth error handling
- Target Quality settings balanced
- Target Quality score calculation fixed and improved
- Default extra_splits set to 240 frames.
- Extraction and Concatenation to copy all streams.
- Scenes file to save total frame count, faster restart/start up with scenes file.
- Fix concat on windows with ffmpeg.
- Revorked args to classes.
- Skip files in queue if they already encoded.
- Default chunk method to ffms2.
- Per frame target quality for SVT-AV1
- Added none split method option
- Added webm output
- Refactored VMAF and Target Quality
- VMAF is now separate, ready for import module
- Changed Target Quality probing rate
- Refactored module structure/names

### 3
- Default pix format to be yuv420p10le
- Default scene change interval to be 120 frames

### 2
- Target VMAF renamed to Target Quality
- Changed Algo for Target Quality score calculation

av1an/arg_parse.py
#!/bin/env python
import argparse
import sys
from pathlib import Path

from .project import Project


class Args:
    """
    Class responsible for arg parsing
    Creation of original project file
    """

    def __init__(self):
        self.parser = self.arg_parsing()
        self.defaults = self.get_defaults()
        self.parsed = None
        self.project = None

    def get_project(self):
        """
        Create and return project object with all parameters
        """
        if not self.parsed:
            self.parse()

        self.project = Project(self.parsed)

        if self.project.config:
            self.save_load_project_file()

        return self.project

    def get_defaults(self) -> dict:
        """
        Get dictionary of default values specified in arg_parsing()
        """
        return vars(self.parser.parse_args([]))

    def get_difference(self) -> dict:
        """
        Return difference of defaults and new
        """
        return dict([x for x in self.parsed.items() if x not in self.defaults.items()])

    def parse(self):
        """
        Parse command line parameters provided by user
        """
        self.parsed = vars(self.parser.parse_args())
        if not self.parsed["input"]:
            self.parser.print_help()
            sys.exit()

    def save_load_project_file(self):
        """
        Saves current/Loads given project file, loads saved project first and when overwrites only unique values from current parse
        """
        cfg_path = Path(self.project.config)

        if cfg_path.exists():

            new = self.get_difference()
            self.project.load_project_from_file(self.project.config)
            self.project.load_project(new)

        else:
            self.project.save_project_to_file(self.project.config)

    def arg_parsing(self):
        """Command line parsing and setting default variables"""
        parser = argparse.ArgumentParser()

        # Input/Output/Temp
        io_group = parser.add_argument_group("Input and Output")
        io_group.add_argument(
            "--input",
            "-i",
            nargs="+",
            type=Path,
            help="Input File or Vapoursynth (.py,.vpy) script",
        )
        io_group.add_argument(
            "--temp", type=str, default=None, help="Set temp folder path"
        )
        io_group.add_argument(
            "--output_file", "-o", type=str, default=None, help="Specify output file"
        )
        io_group.add_argument(
            "--mkvmerge",
            help="Use mkvmerge instead of ffmpeg to concatenate",
            action="store_true",
        )

        io_group.add_argument(
            "--quiet",
            "-q",
            help="Disable printing tqdm and scenedetect data to terminal",
            action="store_true",
        )
        io_group.add_argument(
            "--logging", "-log", type=str, default=None, help="Enable logging"
        )
        io_group.add_argument(
            "--resume", "-r", help="Resuming previous session", action="store_true"
        )
        io_group.add_argument(
            "--keep", help="Keep temporally folder after encode", action="store_true"
        )
        io_group.add_argument(
            "--config",
            "-c",
            type=str,
            default=None,
            help="Path to config file, create if doesn't exists",
        )
        io_group.add_argument(
            "--webm", default=False, action="store_true", help="Output to webm"
        )

        # Splitting
        split_group = parser.add_argument_group("Splitting")
        split_group.add_argument(
            "--chunk_method",
            "-cm",
            type=str,
            default=None,
            help="Method for creating chunks",
            choices=["select", "vs_ffms2", "vs_lsmash", "hybrid"],
        )
        split_group.add_argument(
            "--scenes", "-s", type=str, default=None, help="File location for scenes"
        )
        split_group.add_argument(
            "--split_method",
            type=str,
            default="pyscene",
            help="Specify splitting method",
            choices=["none", "pyscene", "aom_keyframes", "ffmpeg"],
        )
        split_group.add_argument(
            "--extra_split",
            "-xs",
            type=int,
            default=240,
            help="Number of frames after which make split",
        )

        # PySceneDetect split
        split_group.add_argument(
            "--threshold", "-tr", type=float, default=35, help="PySceneDetect Threshold"
        )
        split_group.add_argument(
            "--min_scene_len",
            type=int,
            default=60,
            help="Minimum number of frames in a split",
        )

        # AOM Keyframe split
        split_group.add_argument(
            "--reuse_first_pass",
            help="Reuse the first pass from aom_keyframes split on the chunks",
            action="store_true",
        )

        # Encoding
        encode_group = parser.add_argument_group("Encoding")
        encode_group.add_argument(
            "--passes",
            "-p",
            type=int,
            default=None,
            help="Specify encoding passes",
            choices=[1, 2],
        )
        encode_group.add_argument(
            "--video_params", "-v", type=str, default=None, help="encoding settings"
        )
        encode_group.add_argument(
            "--encoder",
            "-enc",
            type=str,
            default="aom",
            help="Choosing encoder",
            choices=[
                "aom",
                "svt_av1",
                "svt_vp9",
                "rav1e",
                "vpx",
                "x265",
                "x264",
                # "vvc",
            ],
        )
        encode_group.add_argument(
            "--workers", "-w", type=int, default=0, help="Number of workers"
        )
        encode_group.add_argument(
            "--no_check", "-n", help="Do not check encodings", action="store_true"
        )
        encode_group.add_argument(
            "--force",
            help="Force encoding if input args seen as invalid",
            action="store_true",
        )

        # FFmpeg params
        ffmpeg_group = parser.add_argument_group("FFmpeg")
        ffmpeg_group.add_argument(
            "--ffmpeg", "-ff", type=str, default="", help="FFmpeg commands"
        )
        ffmpeg_group.add_argument(
            "--audio_params",
            "-a",
            type=str,
            default="-c:a copy",
            help="FFmpeg audio settings",
        )
        ffmpeg_group.add_argument(
            "--pix_format",
            "-fmt",
            type=str,
            default="yuv420p10le",
            help="FFmpeg pixel format",
        )

        # Vmaf
        vmaf_group = parser.add_argument_group("VMAF")
        vmaf_group.add_argument(
            "--vmaf", help="Calculating vmaf after encode", action="store_true"
        )
        vmaf_group.add_argument(
            "--vmaf_path", type=Path, default=None, help="Path to vmaf models"
        )
        vmaf_group.add_argument(
            "--vmaf_res",
            type=str,
            default="1920x1080",
            help="Resolution used in vmaf calculation",
        )
        vmaf_group.add_argument(
            "--n_threads", type=int, default=None, help="Threads for vmaf calculation"
        )

        # Target Quality
        tq_group = parser.add_argument_group("Target Quality")
        tq_group.add_argument("--target_quality", type=float, help="Value to target")
        tq_group.add_argument(
            "--target_quality_method",
            type=str,
            default="per_shot",
            help="Method selection for target quality",
            choices=["per_frame", "per_shot"],
        )
        tq_group.add_argument(
            "--probes",
            type=int,
            default=4,
            help="Number of probes to make for target_quality",
        )
        tq_group.add_argument(
            "--min_q", type=int, default=None, help="Min q for target_quality"
        )
        tq_group.add_argument(
            "--max_q", type=int, default=None, help="Max q for target_quality"
        )
        tq_group.add_argument(
            "--vmaf_plots",
            help="Make plots of probes in temp folder",
            action="store_true",
        )
        tq_group.add_argument(
            "--probing_rate",
            type=int,
            default=4,
            help="Framerate for probes, 0 - original",
        )
        tq_group.add_argument(
            "--vmaf_filter",
            type=str,
            default=None,
            help="Filter applied to source at vmaf calcualation, use if you crop source",
        )

        # Misc
        misc_group = parser.add_argument_group("Misc")
        misc_group.add_argument(
            "--version", action="version", version=f"Av1an version: 6.1.5"
        )
        misc_group.add_argument(
            "--priority",
            choices=['idle', 'low', 'normal', 'high'],
            default='normal',
            help='Sets process priority (Default: normal)'
        )
        # Initialize project with initial values
        return parser
